### ISSUE TITLE
[GB] read from memory crashes after `reboot_core()`

### ISSUE THREAD

**Issue by eqy** [TASVideos/BizHawk#1716]
I'm a newbie to BizHawk and GB TAS so please forgive me if this is obvious.
I'm inspecting some addresses in Pokemon Blue via a lua script. Reading a byte seems to work the first time, but after calling `reboot_core()`, reading the same byte will crash BizHawk.

This snippet should reproduce the issue:
```
client.speedmode(500)
function n_frames(n)
        for i=1,n do
                emu.frameadvance()
        end
end
n_frames(2000)
client.reboot_core()
n_frames(2000)
val = memory.readbyte(tonumber('0xD18C'))
print("val: " .. tostring(val))
n_frames(2000)
client.reboot_core()
n_frames(2000)
val = memory.readbyte(tonumber('0xD18C'))
print("val: " .. tostring(val))

```

The address should be valid in Pokemon, and corresponds to the level of the first Pokemon.
Interestingly, manually rebooting instead of calling `reboot_core` seems to be ok.

It seems that after the `reload_core()`, the default memory domain is changed to `System Bus`, but after switching this manually back to `WRAM`, the "segment"? is no longer correct---`0xD18C` is interpreted literally as 53,644 instead of `0xD18C - 0xC000`, outside the size of 8,192.

**Comment by alyosha-tas** [comment]
This is not a memory domain issue. Both times it is reading properly from system bus.

The problem is that lua is not using an updated core instance after the reboot. When the first instance gets disposed, the Gambattestate pointer becomes 0. It still uses this to try to access gambatte memory the second time, getting an access violation. You can see this by logging the pointer value when the memory domain is accessed. The first time it uses the correct pointer, the second it uses 0.

How do you fix this? I don't know either.

**Comment by eqy** [comment]
Are there any other technical advantages to using GBHawk (speed, etc?)

A workaround I've found is to just use `mainmemory` and to convert to flat addresses.
e.g., https://github.com/eqy/GeneticAlgorithmPlaysPok-mon/blob/b13307b6adaf427a55ee51a1fa872032b19da6d9/main.lua#L117

**Comment by alyosha-tas** [comment]
nope, its actually much slower then gambatte. If all you are doing is pokemon, gambatte is the way to go. GBHawk is slightly more accurate otherwise.

**Comment by alyosha-tas** [comment]
I tried debugging this a bit today. I'll leave some notes.
I'm pretty sure the problem is here in luaconsole.cs ~ line 170;

```
if (LuaImp.IsRebootingCore)
				{
					// Even if the lua console is self-rebooting from client.reboot_core() we still want to re-inject dependencies
					LuaImp.Restart(Emulator.ServiceProvider);
					
					return;
				}
```

The problem here is that this doesn't make new instances of the 'Libraries' used by LuaImp. These libraries are made only once, and they contain all the information from the core, which gets delegated to an APIs instance somewhere along the way.

So when you 'restart' it doesn't get updated core information.

Currently 'Libraries' is protected readonly, it will probably need to be made public in order to access and update them. Not sure right now how to do it.

**Comment by YoshiRulz** [comment]
I think I fixed this with cf901753b.
